### PR TITLE
http conn man: default preserve_upstream_date to true, add runtime guard

### PR DESCRIPTION
**Pull request by chradcliffe**
Signed-off-by: Craig Radcliffe <craig.radcliffe@broadcom.com>

Commit Message: http conn man: always preserve upstream `date` response header

Additional Description:
* Reintroduces the change to preserve the upstream `date` response header (introduced in #11077, reverted in #11116 ) but removes the configuration and adds a runtime guard instead (see #11110 )

Risk Level: Low

Testing: Unit testing

Docs Changes: N/A

Release Notes: Unsure -- runtime guard introduced

Runtime guard: http_connection_manager.preserve_upstream_date

### CODE REVIEW

File: api/envoy/extensions/filters/network/http_connection_manager/v3/http_connection_manager.proto [hunk 1]
```diff
@@ -507,6 +507,11 @@ message HttpConnectionManager {
   // 3. Tracing decision (sampled, forced, etc) is set in 14th byte of the UUID.
   RequestIDExtension request_id_extension = 36;
 
+  // If `preserve_upstream_date` is true, the value of the `date` header sent by the upstream
+  // host will not be overwritten by the HTTP Connection Manager. Otherwise, the HTTP Connection
+  // Manager will overwrite this value unconditionally. The default value is true.
+  google.protobuf.BoolValue preserve_upstream_date = 38;
```
Comment (alyssawilk) [inline api/envoy/extensions/filters/network/http_connection_manager/v3/http_connection_manager.proto:513]: Given we believe this is more correct, @mattklein123 WDYT about only having the runtime guard and landing the config knob iff anyone wants to preserve the old behavior?

Alternately if we think it's worth leaving the knob in, we should comment about the runtime guard both here and in docs/root/version_history/current.rst and I think we're good to land.

**Comment by chradcliffe** [comment]
While I was looking to resolve the merge conflict caused by the landing of #11100 I just recalled this section in the [RFC](https://tools.ietf.org/html/rfc7231#section-7.1.1.2):

>  A recipient with a clock that receives a response message without a Date header field MUST record the time it was received and append a corresponding Date header field to the message's header section if it is cached or forwarded downstream.

[The logic](https://github.com/envoyproxy/envoy/blob/bb74a91ac413d81f4b1874ed6474df93e4ca2782/source/common/http/conn_manager_impl.cc#L1643) that prevents the date header from being written when the response has been cached doesn't consider the case of the date header not being set. Should we always set the date header if it is not already set, even if the response comes from the cache?

I'm not that familiar with the caching code, so it's possible that the date header is always set for responses that have been cached.

**Comment by alyssawilk** [comment]
I'd be inclined to add the header if not present but pinging @toddmgreer for all things cache related :-)

**Comment by toddmgreer** [comment]
Yes, if there's no date header, we should add one.

**Comment by chradcliffe** [comment]
In case it's not clear, in the latest merge commit I resolved the conflict and modified the logic so that the date header is set when there is no date header in the upstream response, even in the cached-response case.

File: api/envoy/extensions/filters/network/http_connection_manager/v3/http_connection_manager.proto (hunk 1 above)
Comment (mattklein123) [reply api/envoy/extensions/filters/network/http_connection_manager/v3/http_connection_manager.proto:513]: +1 let's just remove the config, have it controlled by runtime, and then we can add config if anyone complains. I find it *highly* unlikely anyone will notice.

File: source/common/http/conn_manager_impl.cc [hunk 2]
```diff
@@ -1640,9 +1640,18 @@ void ConnectionManagerImpl::ActiveStream::encodeHeaders(ActiveStreamEncoderFilte
 void ConnectionManagerImpl::ActiveStream::encodeHeadersInternal(ResponseHeaderMap& headers,
                                                                 bool end_stream) {
   // Base headers.
-  if (!stream_info_.hasResponseFlag(StreamInfo::ResponseFlag::ResponseFromCacheFilter)) {
+
+  // By default, always preserve the upstream date response header if present. If we choose to
+  // overwrite the upstream date unconditionally (a previous behavior), only do so if the response
+  // is not from cache
+  const bool should_preserve_upstream_date =
```
Comment (alyssawilk) [inline source/common/http/conn_manager_impl.cc:1647]: I'd suggest using
Runtime::runtimeFeatureEnabled("")

File: source/common/http/conn_manager_impl.cc [hunk 3]
```diff
@@ -1640,9 +1640,18 @@ void ConnectionManagerImpl::ActiveStream::encodeHeaders(ActiveStreamEncoderFilte
 void ConnectionManagerImpl::ActiveStream::encodeHeadersInternal(ResponseHeaderMap& headers,
                                                                 bool end_stream) {
   // Base headers.
-  if (!stream_info_.hasResponseFlag(StreamInfo::ResponseFlag::ResponseFromCacheFilter)) {
+
```
Comment (alyssawilk) [inline source/common/http/conn_manager_impl.cc:1643]: Please also add a release note of the change, including the runtime guard, to docs/root/version_history/current.rst

File: source/common/http/conn_manager_impl.cc (hunk 2 above)
Comment (chradcliffe) [reply source/common/http/conn_manager_impl.cc:1647]: Should this guard be under the `envoy.reloadable_features` namespace, then?

Comment (mattklein123) [reply source/common/http/conn_manager_impl.cc:1647]: Yeah, this is something we will remove if no one complains.

**Pull request merged**